apollo-federation/src/sources/connect/json_selection/apply_to.rs

```diff
@@ -242,53 +240,79 @@ impl ApplyTo for NamedSelection {
     }
 }
 
-// $typenames is a special variable for referring to literal typenames. See
-// note in selection_set.rs for more detail.
-pub(super) const TYPENAMES: &str = "$typenames";
-
 impl ApplyTo for PathSelection {
     fn apply_to_path(
         &self,
         data: &JSON,
-        vars: &IndexMap<String, JSON>,
-        input_path: &mut Vec<JSON>,
+        vars: &VarsWithPathsMap,
+        input_path: &InputPath<JSON>,
         errors: &mut IndexSet<ApplyToError>,
     ) -> Option<JSON> {
-        if let JSON::Array(array) = data {
-            return self.apply_to_array(array, vars, input_path, errors);
-        }
-
-        match self {
-            Self::Var(var_name, tail) => {
-                if var_name == "$" {
-                    // Because $ refers to the current value, we keep using
-                    // input_path instead of creating a new var_path here.
-                    tail.apply_to_path(data, vars, input_path, errors)
-                } else if var_name == TYPENAMES {
-                    if let PathSelection::Key(Key::Field(ref name), _) = **tail {
-                        let var_data = json!({ name: name });
-                        let mut var_path = vec![json!(name)];
-                        tail.apply_to_path(&var_data, vars, &mut var_path, errors)
+        match &self.path {
+            // If this is a KeyPath, instead of using data as given, we need to
+            // evaluate the path starting from the current value of $. To
+            // evaluate the KeyPath against data, prefix it with @. This logic
+            // supports method chaining like obj->has('a')->and(obj->has('b')),
+            // where both obj references are interpreted as $.obj.
+            PathList::Key(key, tail) => {
+                if let Some((dollar_data, dollar_path)) = vars.get("$") {
+                    let input_path_with_key = dollar_path.append(key.to_json());
+                    if let Some(child) = dollar_data.get(key.as_string()) {
+                        tail.apply_to_path(child, vars, &input_path_with_key, errors)
                     } else {
                         errors.insert(ApplyToError::new(
-                            format!("Invalid {} usage", TYPENAMES).as_str(),
-                            &[json!(var_name), json!(tail)],
+                            format!(
+                                "Property {} not found in {}",
+                                key.dotted(),
+                                json_type_name(dollar_data),
+                            )
+                            .as_str(),
```

Comment:
```suggestion
```

Again, this could be saved for a follow-up since there are probably a ton of them. I'll stop highlighting them now.

Reply:
https://github.com/apollographql/router/pull/5762/commits/d73bb4d15aa2ce22bd251a4b3746f22262b27571